src/data_tools/src/jsf_data.cpp

```diff
@@ -0,0 +1,280 @@
+ */
+
+#include <data_tools/jsf_data.h>
+
+#define BOOST_NO_CXX11_SCOPED_ENUMS
+#include <boost/date_time.hpp>
+#include <boost/date_time/gregorian/gregorian.hpp>
+#undef BOOST_NO_CXX11_SCOPED_ENUMS
+
+#include <fcntl.h>
+#ifdef _MSC_VER
+  #include <io.h>
+#else
+  #include <unistd.h>
+#endif
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+
+# define SONAR_MESSAGE_HEADER_START 0X1601
+# define SONAR_DATA_TYPE 0X0050
+# define DVL_DATA_TYPE 0X0820
+# define SITUATION_DATA_TYPE 0X082A
+
+
+using namespace std;
+
+namespace jsf_data{
+
+
+cv::Mat make_waterfall_image(const jsf_sss_ping::PingsT& pings)
+{
+    int rows = pings.size();
+    int cols = pings[0].port.pings.size() + pings[0].stbd.pings.size();
+    cv::Mat swath_img = cv::Mat(rows, cols, CV_8UC3, cv::Scalar(255, 255, 255));
+    for (int i = 0; i < pings.size(); ++i) {
+        for (int j = 0; j < pings[i].port.pings.size(); ++j) {
+            cv::Point3_<uchar>* p = swath_img.ptr<cv::Point3_<uchar> >(i, pings[0].stbd.pings.size()+j);
+            p->z = uchar(255.*pings[i].port.pings[j]);
+            p->y = uchar(255.*pings[i].port.pings[j]);
+            p->x = uchar(255.*pings[i].port.pings[j]);
+        }
+        for (int j = 0; j < pings[i].stbd.pings.size(); ++j) {
+            cv::Point3_<uchar>* p = swath_img.ptr<cv::Point3_<uchar> >(i, pings[0].stbd.pings.size()-j-1);
+            p->z = uchar(255.*pings[i].stbd.pings[j]);
+            p->y = uchar(255.*pings[i].stbd.pings[j]);
+            p->x = uchar(255.*pings[i].stbd.pings[j]);
+        }
+    }
+    cv::Mat resized_swath_img;//dst image
+    cv::resize(swath_img, resized_swath_img, cv::Size(cols/32, rows/4));//resize image
+    
+    return resized_swath_img;
+}
+
+
+
+
+void show_waterfall_image(const jsf_sss_ping::PingsT& pings)
+{
+    cv::Mat waterfall_image = make_waterfall_image(pings);
+    cv::imshow("Waterfall image", waterfall_image);
+    cv::waitKey();
+}
+
+// skip data other than sonar data (type=80)
+void skip_data(ifstream& input, jsf_msg_header jsf_hdr){
+    int nbr_bytes;
+    for (int i = 0; i < jsf_hdr.following_bytes; i++){   
+        input.read(reinterpret_cast<char*>(&nbr_bytes), sizeof(char));
+    }
+}
+
+
+jsf_sss_ping_side process_side_scan_ping_side(ifstream& input, jsf_msg_header& jsf_hdr, jsf_sonar_data_msg_header& jsf_sonar_data_hdr){
+    jsf_sss_ping_side ping_side;
+    if(jsf_hdr.start_marker!=SONAR_MESSAGE_HEADER_START){
+        cout<<"error start marker: "<<jsf_hdr.start_marker<<endl;
+        exit(0);
+    }
+    if(jsf_hdr.msg_type!= SONAR_DATA_TYPE){
+        cout<<"Error message type: "<<jsf_hdr.msg_type<<endl;
+        exit(0);
+    }
+
+    if (jsf_sonar_data_hdr.data_format==0){
+        int16_t env_data;
+        float scaled_env_data[jsf_sonar_data_hdr.spls_num_in_pkt];
```

Comment:
This is the line that fails. I think it is not a constant so this should indeed not compile? You need to use something like `float* scaled_env_data = new float[jsf_sonar_data_hdr.spls_num_in_pkt]`.

Reply:
Or maybe just use a vector: `vector<float> scaled_env_data(jsf_sonar_data_hdr.spls_num_in_pkt)`.

Reply:
That would be my preference.

Reply:
Thanks for the suggestions! I guess we have to try these to see if it can work. I have changed it to float* scaled_env_data = new float[jsf_sonar_data_hdr.spls_num_in_pkt]. Let's see if it can pass to compile.